pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -366,8 +398,47 @@ ssize_t sock_dtls_send(sock_dtls_t *sock, sock_dtls_session_t *remote,
         }
     }
 
-    return dtls_write(sock->dtls_ctx, &remote->dtls_session,
-                      (uint8_t *)data, len);
+    res = dtls_write(sock->dtls_ctx, &remote->dtls_session,
+                     (uint8_t *)data, len);
+#ifdef SOCK_HAS_ASYNC
+    if ((res >= 0) && (sock->async_cb != NULL)) {
+        sock->async_cb(sock, SOCK_ASYNC_MSG_SENT, sock->async_cb_arg);
+    }
+#endif /* SOCK_HAS_ASYNC */
+    return res;
+}
+
+#if SOCK_HAS_ASYNC
+/**
+ * @brief   Checks for and iterates for more data chunks within the network
+ *          stacks anternal packet buffer
+ *
+ * When no more chunks exists, `data_ctx` assures cleaning up the internal
+ * buffer state and `sock_udp_recv_buf()` returns 0.
+ *
+ * @see @ref sock_udp_recv_buf().
+ */
+static void _check_more_chunks(sock_udp_t *udp_sock, void **data,
+                               void **data_ctx, sock_udp_ep_t *remote)
+{
+    ssize_t res;
+
+    while ((res = sock_udp_recv_buf(udp_sock, data, data_ctx, 0, remote)) > 0) {
```

Comment:
I'm currently trying to implement `sock_dtls_recv_buf_aux()` so I got back to this function.
How is this supposed to work? The way this is implemented (the while loop) this will just flush (and discard) pending data on the socket.

On tangent: Is there even such a thing as 'chunked datagram payload'?
As far as I can see it is implemented in no network stack, it's always

 - first call returns pointer to the buffer
 - second call frees the buffer

The only way I could see we could get chunked datagram payload is if we didn't do IP fragment reassembly in the stack but pushed it to the user to handle.

That seems not like something anyone would want. And user would need to maintain a reassembly buffer, nullifying the zero-copy advantage of `sock_udp_recv_buf()`.

Is there any real world use for this API that we can't do a single buffer lease -> return?

Reply:
> On tangent: Is there even such a thing as 'chunked datagram payload'?

Yes, lwIP is returning ~500B---I don't remember the exact number and don't want to stall answering this question by looking it up---non-continuous chunks of data for larger UDP packets. Not sure if this is still the case, after the packet went to TinyDTLS or if other DTLS implementations might still opt to chunk the payload (e.g. by data record in the actual DTLS packet).

> How is this supposed to work?

Due to the restrictions of TinyDTLS that you I believe fixed a proper implementation was not possible so far. The idea is to call this function repeatedly, until there are no chunks left. I think the `lwip_sock_udp` implementation provides a good example of how this is supposed to look like.